posts/inside-rust/2023-07-21-crates-io-postmortem.md

```diff
@@ -0,0 +1,178 @@
+
+    - The pull request introduced tests, but did not test all code paths.
+
+      **Why did the pull request not test all code paths?**
+
+      - The code was structured in a way that made testing with different "CDN prefix" values complicated.
+
+        **Why was the code structured in a way that made testing different values complicated?**
+
+        - The code had not been unit tested before and the refactoring stopped at a point where the code could at least be tested with a hardcoded value.
+
+          **Why did the refactoring stop at that point?**
+
+          - It was deemed "good enough for now" by the developer.
+
+    - The pull request was not reviewed by another developer.
+
+      **Why was the pull request not reviewed by another developer?**
+
+      - The developer creating the pull request misjudged the potential impact of a bug in the pull request. They did not explicitly request a review from the crates.io team and merged it themselves after a few hours.
+
+        **Why was no code review requested from the crates.io team?**
+
+        - The number of active team members in the [crates.io team](https://www.rust-lang.org/governance/teams/crates-io) is quite small. Reviewing dozens of pull requests per months from the one developer who is employed to work fulltime on crates.io would be a recipe for burnout for the other members of the crates.io team. For that employed fulltime developer it would also not work well if they were blocked on waiting for reviews for the majority of their time. The current way of working is that code reviews are only requested for high-impact pull requests. 
+
+        **Why was the potential impact misjudged?**
+
+        - The developer forgot to think about the fact that this change affected the crate download endpoint of crates.io, which is the endpoint that handles 99% of the traffic to the server.
+
+          **Why did the developer forget to check if a high-priority endpoint is affected?**
+
+          - There is no checklist or guide describing in which case a pull request should be seen as having a high potential impact and thus needing explicit code review from the crates.io team.
+
+    **Why did the bug make it into production?**
+
+    - The crate download endpoint was not tested on the staging environment before promoting it to production.
+
+      **Why was the crate download not tested?**
+
+      - The test plan for the staging environment only includes publishing a new version and seeing that reflected on the website and in the package index repository.
+
+        **Why does the test plan not include crate downloads?**
+
+        - Since 99% of all requests to crates.io are for crate downloads, the test plan definitely should include this process. There is intentionally no download button on the webpage though, so the URL for the download has to be constructed manually.
+  
+          **Why does the download URL need to be constructed manually?**
```

Comment:
I mean, part of the test plan could be "add the crate you just published as a dependency to another crate and see that it builds successfully", which would be less annoying than downloading the URL manually.

Reply:
right, that would be another option.